### PR TITLE
Fix Variable and Secret url bugs

### PR DESCRIPTION
**Pull request by AndrewJDawes**
Fixes: #2815, #2834

Tests added to `tests/Repository.py`:

- `testRepoSecrets`
- `testRepoVariables`

Note:

- This fix for Variable and Secret urls is a dependency for new functionality to fetch `Environment` variables and secrets
  - i.e., [https://docs.github.com/en/rest/actions/variables?apiVersion=2022-11-28#list-environment-variables](https://docs.github.com/en/rest/actions/variables?apiVersion=2022-11-28#list-environment-variables)
- Adding the optional `attributeTransformer` callable argument will also make it possible to support this Variable and Secret functionality on the `Environment` model.
- I am currently working on a pull request to add support for `get_secrets` and `get_variables` on the `Environment` model (tests are not added yet)!

**If you have any questions or if there is anything I can help with, would you please let me know?**

We utilize this library at my work place to for automations against repos in our GitHub Enterprise organizations, so we are eager to get some of these bugs fixed and new functionality added to fully support Environment variables and secrets.

### CODE REVIEW

**Comment by AndrewJDawes** [comment]
Also fixes #2785

**Comment by EnricoMi** [comment]
Thanks for the contribution.

I'd prefer to move the logic that builds the Variable url from the Repository (that uses an attributes transformer), into the Variable class. 

For this:
- add attribute `variables_url` to Variable
- in `Variable.url`, construct `url` from `variables_url` and `name`, if `self._url.` is not set
- in `Repository`, inject `variables_url` into paginated elements via `attributes_transformer_override_from_dictionary`

**Comment by EnricoMi** [comment]
Btw, in 1b85bfc I have restored compatibility of `PaginatedList` with Pickle, which is good to have.

**Comment by AndrewJDawes** [comment]
Hey, @EnricoMi !

Pushed up a new commit to implement your suggestions!

**Comment by EnricoMi** [comment]
> Pushed up a new commit to implement your suggestions!

Looks exactly like what I meant! Now lets iterate over the PR some more...

File: github/PaginatedList.py [hunk 1]
```diff
@@ -153,12 +154,18 @@ def __init__(
...
 
         first_page = []
         if firstData is not None and firstHeaders is not None:
             first_page = self._getPage(firstData, firstHeaders)
         super().__init__(first_page)
 
+    def _transformAttribute(self, attribute: Any) -> Any:
+        if self._attributesTransformer is None:
+            return attribute
+        return self._attributesTransformer(attribute)
```
Comment (EnricoMi) [inline github/PaginatedList.py:167]: I got the naming / semantics wrong, I think, this is more accurate:
```suggestion
    def _transformAttributes(self, element: Dict[str, Any]) -> Dict[str, Any]:
        if self._attributesTransformer is None:
            return element
        return self._attributesTransformer(element)
```

File: github/PaginatedList.py [hunk 2]
```diff
@@ -270,5 +276,29 @@ def get_page(self, page: int) -> List[T]:
...
+            self.__contentClass(self.__requester, headers, self._transformAttribute(element), completed=False)
+            for element in data
+        ]
 
-        return [self.__contentClass(self.__requester, headers, element, completed=False) for element in data]
+    @classmethod
+    def attributes_transformer_override_from_dictionary(
+        cls, transformer: Callable[[Dict[str, Any]], Dict[str, Any]], overrides: dict
+    ) -> Callable[[Dict[str, Any]], Dict[str, Any]]:
+        def attributes_transformer(attributes: Dict[str, Any]) -> Dict[str, Any]:
```
Comment (EnricoMi) [inline github/PaginatedList.py:288]:
```suggestion
        def attributes_transformer(element: Dict[str, Any]) -> Dict[str, Any]:
```

File: github/PaginatedList.py [hunk 3]
```diff
@@ -140,6 +140,7 @@ def __init__(
         list_item: str = "items",
         firstData: Optional[Any] = None,
         firstHeaders: Optional[Dict[str, Union[str, int]]] = None,
+        attributesTransformer: Optional[Callable[[Any], Any]] = None,
```
Comment (EnricoMi) [inline github/PaginatedList.py:143]:
```suggestion
        attributesTransformer: Optional[Callable[[Dict[str, Any]], Dict[str, Any]]] = None,
```

File: github/PaginatedList.py [hunk 4]
```diff
@@ -270,5 +276,29 @@ def get_page(self, page: int) -> List[T]:
...
             data = data[self.__list_item]
+        return [
+            self.__contentClass(self.__requester, headers, self._transformAttribute(element), completed=False)
+            for element in data
+        ]
 
-        return [self.__contentClass(self.__requester, headers, element, completed=False) for element in data]
+    @classmethod
+    def attributes_transformer_override_from_dictionary(
+        cls, transformer: Callable[[Dict[str, Any]], Dict[str, Any]], overrides: dict
```
Comment (EnricoMi) [inline github/PaginatedList.py:286]:
```suggestion
        cls, transformer: Callable[[Dict[str, Any]], Dict[str, Any]], overrides: Dict[str, Any]
```

File: github/PaginatedList.py [hunk 5]
```diff
@@ -270,5 +276,29 @@ def get_page(self, page: int) -> List[T]:
...
+    ) -> Callable[[Dict[str, Any]], Dict[str, Any]]:
```
Comment (EnricoMi) [inline github/PaginatedList.py:287]: Why do we need this extra `transformer`? Can we simplify this to

```suggestion
    def override_attributes(cls, overrides: Dict[str, Any]) -> Callable[[Dict[str, Any]], Dict[str, Any]]:
```

and then use it as

```python
attributesTransformer=PaginatedList.override_attributes({"secrets_url": f"{self.url}/actions/secrets"}),
```

**Comment by AndrewJDawes** [comment]
Hey, @EnricoMi !

Those suggestions make sense to me! I think it reads cleaner.

I submitted another commit that implements your suggestions. **Would you take a look when you get a chance?**

*Context on why the attributes transformers were initially accepting a `Callable`:*

- I am still hoping to submit a follow-up PR t after this one to provide the same functionality for Repository Environments and their Variables and Secrets.
- To that end, I wanted to make sure there these enhancements wouldn't prevent us from adding that functionality. - Specifically, I wasn't sure about moving control over the attributes transformations into `PaginatedList` and removing the possibility of a decorator pattern to wrap transformers (that was why `lambda` functions and each transformer accepted a `Callable`).
- However, after looking things over, I think moving control to `PaginatedList` and removing the decorator pattern should be fine and we should still be able to add support for Repository Environments and their Variables/Secrets.
- **So, I implemented your suggestions!**

**Comment by AndrewJDawes** [comment]
Hey, @EnricoMi - just noting that I removed the extra unwanted `transformer`. Thought I had done that already - my mistake!

Would you please let me know if there is anything else I can help with?

**Review by EnricoMi — APPROVED** [review]
This looks really good, thanks for the contribution!

**Pull request merged**